Story/CITE-85 Delete multiple citations

SayaliTanawade opened the pull request:
Code clean is not done. I am raising this pull request to get confirmation about my approach for deleting citations.

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented:
code factor failed. as per the pull request guidelines, please address them before requesting a review.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -380,4 +380,18 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+                    citationRepository.delete((Citation)oldCitation.get());
+                } 
+            }
+        } catch (HttpClientErrorException ex) {
```
where is this thrown?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -380,4 +380,18 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
-}
+
+    @Override
+    public void deleteCitations(IUser user, List<String> citationList) throws CannotFindCitationException {
```
where is the user used?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/zotero/impl/ZoteroManager.java:
```diff
@@ -37,6 +40,7 @@
+import edu.asu.diging.citesphere.model.bib.impl.Citation;
 import edu.asu.diging.citesphere.model.bib.impl.CitationCollectionResult;
 import edu.asu.diging.citesphere.model.bib.impl.CitationResults;
 import edu.asu.diging.citesphere.user.IUser;
```
why are all these imports added if there are no code changes?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,67 @@
+    @RequestMapping(value = "/auth/group/{groupId}/delete", method = RequestMethod.POST)
+    public String delete(Authentication authentication,@PathVariable("groupId") String groupId, @RequestParam(value="citationList", required=false) String citationList)
+            throws ZoteroConnectionException, GroupDoesNotExistException, ZoteroHttpStatusException, ZoteroItemCreationFailedException {
+        String [] list = citationList.split(",");
```
where is this used?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,67 @@
+        try {
+            citationManager.deleteCitations((IUser) authentication.getPrincipal(), citations);
+        } catch (CannotFindCitationException e) {
+            e.printStackTrace();
```
needs to be properly handled

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,67 @@
+        }
+        
+
+        return "redirect:/auth/group/"+groupId+"/items";
```
I believe you can just do `.../group/{groupId}/items`

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -64,6 +66,9 @@
     private CitationGroupRepository groupRepository;
+    
+    @Autowired
+    private CitationRepository citationRepository;
```
please use citationstore

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -379,4 +384,14 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+        for(String citation : citationList) {
+            Optional<ICitation> oldCitation = citationRepository.findByKey(citation);
+            if(oldCitation.isPresent()) {
+                citationRepository.delete((Citation)oldCitation.get());
```
use the citationstore as the indexer needs to be updated when deleting citations

jdamerow replied:
also, you need to delete the citations from zotero, don't you?

jdamerow replied:
and when you do, make sure to make one request to zotero with all citations to be deleted not multiple (i'm fairly certain that's possible).

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,56 @@
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Value("${_creation_default_item_type}")
+    private String defaultItemType;
```
why is that here?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Autowired
+    @Qualifier("creatorsFile")
+    private Properties properties;
```
and this?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,56 @@
+
+
+    @RequestMapping(value = "/auth/group/{groupId}/delete", method = RequestMethod.POST)
+    public String delete(Authentication authentication,@PathVariable("groupId") String groupId, @RequestParam(value="citationList", required=false) String citationList)
```
I think you can have spring parse the list of ids directly into a list or array by not using `String citationList` (but List or array).

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,56 @@
+    private ICitationManager citationManager;
+
+
+    @RequestMapping(value = "/auth/group/{groupId}/delete", method = RequestMethod.POST)
```
let's make this `/auth/group/{groupId}/references/delete`

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -379,4 +384,14 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
-}
+
+    @Override
+    public void deleteCitations(List<String> citationList) throws CannotFindCitationException {
```
so, if one citation isn't found, the rest is not getting deleted? It should delete all other citations even if one can't be found.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,56 @@
+            citationManager.deleteCitations(citations);
+        } catch (CannotFindCitationException e) {
+            logger.error("Cition not found.", e);
+            return new String ("{\"error\": \"Selected citation(s) not found.\"}" +  HttpStatus.NOT_FOUND);
```
see my comment in `citationManager`

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -334,6 +409,42 @@ <h4 class="panel-title">
+        <h4 class="modal-title">Delete Entry</h4>
+      </div>
+      <div class="modal-body">
+        <p>Please select at least 1 Entry !!!</p>
```
This doesn't need three "!", there shouldn't be a blank between Entry and "!", and "Entry" needs to be lowercased.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -379,4 +380,15 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+            Optional<ICitation> citation = citationStore.findById(element);
+            if(citation.isPresent()) {
+                citationStore.delete(citation.get());
+                zoteroManager.deleteCitation(user, groupId, citation.get());
```
let's make one deletion request instead of several. Zotero API allows deleting up to 50 items with one request:
https://www.zotero.org/support/dev/web_api/v3/write_requests#deleting_multiple_items

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,37 @@
+@Controller
+public class DeleteCitationController {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
```
is that used anywhere?

SayaliTanawade commented:
Used Zotero API to delete multiple citations instead of several calls in for loop.

jdamerow commented:
Did you make changes to the zotero plugin? There doesn't seem to be a `deleteAll` method, so I assume so? Please open the PR for the plugin. Also, did you push the code for the zotero plugin? I don't see it. Your code has to be pushed at least at the end of your shift. No exceptions.

SayaliTanawade commented:
I updated zotero plugin and raised PR on spring-social-zotero. I updated method name to deleteMultipleCitations() to make it self-explanatory and pushed the changes here. I have tested the delete functionality using updated zotero plugin and I am able to delete multiple citations.

jdamerow reviewed (changes requested):
I think the changes to the pom are missing?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/zotero/IZoteroConnector.java:
```diff
@@ -67,5 +67,9 @@ void clearCollectionItemsCache(IUser user, String groupId, String collectionId,
+    
+    void deleteItem(IUser user, String groupId, String citationKey, Long citationVersion) throws ZoteroConnectionException, ZoteroHttpStatusException;
+    
+    void deleteMultipleCitations(IUser user, String groupId, String citationKey, Long citationVersion) throws ZoteroConnectionException, ZoteroHttpStatusException;
```
since all the other methods use `Item` instead of `Citation` this one should be too for consistency.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/zotero/IZoteroManager.java:
```diff
@@ -63,5 +63,8 @@ CitationResults getCollectionItems(IUser user, String groupId, String collection
+    
+    void deleteCitation(IUser user, String groupId, String citationKey, Long citationVersion) throws ZoteroConnectionException, ZoteroHttpStatusException;
+    
+    void deleteMutipleCitations(IUser user, String groupId, String citationKey, Long citationVersion) throws ZoteroConnectionException, ZoteroHttpStatusException;
```
same as above, all the other methods seem to use "Item" instead of "citation"

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -379,4 +380,15 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+
+    @Override
+    public ResponseEntity<String> deleteCitations(IUser user, String groupId, List<String> citationList) throws ZoteroConnectionException, ZoteroHttpStatusException {    
+        for(String element : citationList) {
```
`citationList` is a list of ids, right? the name should say that. same in the loop; the loop variable should have a fitting name.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -379,4 +380,15 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+                citationStore.delete(citation.get());
+            }
+        }  
+        return zoteroManager.deleteMultipleItems(user, groupId, citationList, zoteroManager.getLatestGroupVersion(user, groupId));
```
I would first delete the citations from Zotero and then locally. If something goes wrong with Zotero and items were not deleted the data is inconsistent otherwise.

jdamerow commented:
Please reopen when the dependency has a pull request.

SayaliTanawade commented:
I made changes to get the response from spring-social-zotero in the form of List and I am checking it at deleteCitationController

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -99,6 +100,77 @@
+			},
+			'error': function(data) {
+				console.log(data);
+				$.notify('<i class="fas fa-exclamation-circle"></i> Citation could not be deleted!', {
```
"Citations"

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -99,6 +100,77 @@
+		var rows = ids.split(",");
+		var baseurl = "references/delete/";
+
+		$.ajax({
```
how long does the deletion take? If it takes too long, Apache might time out.

SayaliTanawade replied:
This comment is yet to be resolved. I was reading about how to handle ajax calls that take long time. I re-opened PR for your 
review on other changes corresponding to the changes in Spring-social-zotero.

jdamerow replied:
instead of making one call and waiting, there should be an async call to start the delete process and then polling to get the results.

SayaliTanawade commented:
I am re-opening Pull request as I changed code according to changes in Spring-social-zotero. One review comment(how long does the deletion take? If it takes too long, Apache might time out.) is yet to be resolved. I will work on it and update the PR.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -379,4 +380,16 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+        for(String citationId : citationIdList) {
+            Optional<ICitation> citation = citationStore.findById(citationId);
+            if(citation.isPresent()) {
+                citationStore.delete(citation.get());
```
you wanna make sure the deletion was successful before you delete the items locally.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,39 @@
+    @RequestMapping(value = "/auth/group/{groupId}/references/delete", method = RequestMethod.POST)
+    public String delete(Authentication authentication, @PathVariable("groupId") String groupId, @RequestParam(value="citationList", required=false) List<String> citationList)
+            throws ZoteroConnectionException, ZoteroHttpStatusException, GroupDoesNotExistException {
+        List<ItemDeletionResponse> zoteroResponse = new ArrayList<ItemDeletionResponse>();
```
why do you initialize the list if you just replace it anyways in the next line?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,39 @@
+        zoteroResponse = citationManager.deleteCitations((IUser) authentication.getPrincipal(), groupId, citationList);
+        for (ItemDeletionResponse response: zoteroResponse) {
+            if (response.toString() != "SUCCESS") {
+                throw new ZoteroHttpStatusException("Could not delete items. Error: " + response);
```
this will just show an error page to the user, which is not very user-friendly. also, if the first couple got deleted but only the last one didn't, this would also not be accurate. Show an appropriate error message to the user on the items page.

jdamerow replied:
This should return a proper responseentity with a proper error code

jdamerow reviewed (changes requested):
Please make the PR a draft PR if it is not everything has been addressed.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -379,4 +380,27 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+                zoteroManager.getLatestGroupVersion(user, groupId));
+        int counter = 0;
+        for (ItemDeletionResponse res : response) {
+            if (res.toString() == "SUCCESS") {
```
strings need to be compared using `equals`

jdamerow replied:
also, isn't `ItemDeletionResponse` an enum? why don't you compare the enum?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,44 @@
+            @RequestParam(value = "citationList", required = false) List<String> citationList)
+            throws ZoteroConnectionException, ZoteroHttpStatusException, GroupDoesNotExistException {
+        List<ItemDeletionResponse> zoteroResponse = new ArrayList<ItemDeletionResponse>();
+        citationManager.deleteCitations((IUser) authentication.getPrincipal(), groupId, citationList)
```
doesn't `deleteCitations` return a list? why don't you assign that to the above list?

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -99,6 +100,95 @@
+				});
+			},
+			'error': function(data) {
+				console.log(data);
```
remove

jdamerow commented:
@MaulikLim would you please resolve those conflicts and also check on the related pull request in the plugin?

MaulikLim commented:
@jdamerow I have resolved the merge conflicts and fixed a bug on moving multiple citations to a collection. This can be merged after merging and releasing the corresponding plugin for social-spring-zotero 0.11.

jdamerow commented:
Resolve conflicts please

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -438,6 +439,29 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+        int counter = 0;
+        for (ItemDeletionResponse res : response) {
+            if (res.equals(ItemDeletionResponse.SUCCESS)) {
+                for (int i = 0; i < 50; i++) {
```
see my comment on the plugin pr. Once we have the ids for each response, this should loop over the ids rather than over 50. If 50 ever changes, we otherwise have to change not just the plugin but this code as well.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/DeleteCitationController.java:
```diff
@@ -0,0 +1,41 @@
+        List<ItemDeletionResponse> zoteroResponse = citationManager.deleteCitations((IUser) authentication.getPrincipal(), groupId, citationList);
+        for (ItemDeletionResponse response : zoteroResponse) {
+            if (!response.equals(ItemDeletionResponse.SUCCESS)) {
+                throw new ZoteroHttpStatusException("Could not delete items. Error: " + response.getValue() + " " + response.getDescription());
```
this should return an error code

MaulikLim commented:
Made the deletion asynchronous. For some reason, git is showing the difference as removal and insertion of entire class code for some of the files.

jdamerow commented on citesphere/pom.xml:
```diff
@@ -23,7 +23,7 @@
         <javers.version>5.0.0</javers.version>
         <spring.kafka.version>2.2.6.RELEASE</spring.kafka.version>
-        <spring-social-zotero.version>0.10</spring-social-zotero.version>
+        <spring-social-zotero.version>0.11</spring-social-zotero.version>
```
I don't think this release exists?

MaulikLim replied:
Not yet. But there is a corresponding plugin PR.

jdamerow replied:
I think I made comment to move the changes from the plugin into citesphere directly?

jdamerow replied:
Never mind, I think I confused that with the other PR.

jdamerow reviewed (changes requested):
It looks like you're changing the line endings of some files. Please fix and adjust your IDE settings.

MaulikLim commented:
My IDE settings are correct. Not sure why it changed the encoding. Anyway, I have manually changed it back.

MaulikLim commented:
The indentation issue flagged by CodeFactor looks correct.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AsyncCitationManager.java:
```diff
@@ -86,8 +100,66 @@ public AsyncUpdateCitationsResponse getUpdateCitationsResponse(String taskId)
+            if (citation.isPresent()) {
+                citationStore.delete(citation.get());
+            }
+        }
```
removing the citations from the local db shouldn't be done in the clear method. The clear method should just do what it says, clearing the task from the local cache. Especially since this method is called from a controller, what if the user closes the browser before the clear method can ever be called? Return a future implementation that allows you to add a callback method to this.

MaulikLim commented:
Addressed review comment - added callback method to delete citations from local database

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AsyncCitationManager.java:
```diff
@@ -86,8 +94,59 @@ public AsyncUpdateCitationsResponse getUpdateCitationsResponse(String taskId)
+     * 
+     * @param taskId Id of the task
+     */
+    public void clearDeleteTask(String taskId) throws InterruptedException, ExecutionException {
```
I'm going back and forth on this, but I think the better approach is to have a cron job that checks the age of tasks and removes them (say after 24 hours). Otherwise, if the application isn't restarted for a long time and many people just close the browser or navigate away from the page the tasks are never cleared and pile up. If you just add a time stamp and use Spring's cron job feature (might be a new plugin, not sure), it should be fairly straight forward and you can remove the frontend logic to remove tasks.

MaulikLim replied:
Okay. I will take a look.

jdamerow commented:
Also, let's make code factor happy and fix the indentation.

MaulikLim commented:
> Also, let's make code factor happy and fix the indentation.

The indentation is correct, I think. I auto-indented it with Eclipse too. And it is same as before.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AsyncCitationManager.java:
```diff
@@ -16,23 +19,35 @@
  *
  */
 @Component
+@EnableScheduling
```
this should go on one of the config classes

jdamerow replied:
also, it needs test cases

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AsyncCitationManager.java:
```diff
@@ -8,6 +8,9 @@
 import java.util.concurrent.Future;
 
```
I didn't notice that before but this class should not be in the controller package. It's a service class.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AsyncDeleteCitationsResponse.java:
```diff
@@ -0,0 +1,36 @@
+package edu.asu.diging.citesphere.web.user;
```
same as above, unless this class is used exclusively by controllers, it shouldn't be in the controller package.

MaulikLim commented:
The written test cases are for a combination of (update, updateStatus) and (delete, deleteStatus) methods since the returned taskId needs to be used to query back for status. The other approach is to mock the map.get(taskId) methods from the private static final variables of the AsyncCitationManager. But it cannot be done in a simple manner.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/model/async/AsyncDeleteCitationsResponse.java:
```diff
@@ -0,0 +1,38 @@
+package edu.asu.diging.citesphere.core.model.async;
```
The model package should contain the main business logic classes (e.g. citation, person, etc). This class is only used to transfer information from zotero to the front end, isn't it? Therefore I would put it in the package or a subpackage with the corresponding service classes.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/model/async/AsyncUpdateCitationsResponse.java:
```diff
@@ -1,7 +1,9 @@
-package edu.asu.diging.citesphere.web.user;
+package edu.asu.diging.citesphere.core.model.async;
```
same here

jdamerow reviewed (changes requested):
For the tests, let's leave them like they are but for the future using ReflectionTestUtils should let you create the maps as you need them to test each method individually.

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented:
In the modal there is a question mark missing at the end and it should say how many entries will be deleted.  Also, it needs to say "entries" not "Entry".
![image](https://user-images.githubusercontent.com/8881141/142669279-97b60cc4-0227-4606-8574-099f4983543a.png)

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -403,6 +513,42 @@ <h4 class="panel-title">
+        <h4 class="modal-title">Delete Entry</h4>
+      </div>
+      <div class="modal-body">
+        <p>Are you sure you want to delete the selected Entries?</p>
```
"Entries" should be lower case.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -403,6 +513,42 @@ <h4 class="panel-title">
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+        <h4 class="modal-title">Delete Entry</h4>
```
should be "Delete Entries"

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -403,6 +513,42 @@ <h4 class="panel-title">
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+        <h4 class="modal-title">Delete Entry</h4>
```
Should be "Delete Entries"

jdamerow reviewed (changes requested):
Deleting can take some time (right now I was waiting for probably like 5-10 seconds). I thought it was broken because it took so long. There should be an indication that it is doing what it does. Ideally, the entries would disappear from the table and a spinning icon would indicate that it is doing something.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/AsyncCitationManager.java:
```diff
@@ -0,0 +1,166 @@
+    private final Map<String, Future<Map<ItemDeletionResponse, List<String>>>> deleteTaskTracker;
+    private final Map<String, Long> deleteTaskTimestamps;
+    
+    private static final long durationInMillis = 86400000L; //Number of milliseconds in a day
```
We probably want this to be configurable

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/AsyncCitationManager.java:
```diff
@@ -0,0 +1,166 @@
+ *
+ */
+@Component
+public class AsyncCitationManager {
```
this class needs an interface

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/items.html:
```diff
@@ -264,6 +376,7 @@ <h4 class="panel-title">
 </div>
 <div class="pull-right" style="margin-top: 20px;">
 <a th:href="@{|/auth/group/${zoteroGroupId}/items/create|}" class="btn btn-primary"><i class="icon-circle-add"></i> Create Citation</a>
+<a class="btn btn-primary delete-link">Delete Citation</a>
```
Should be "Delete Citations" with a trash icon

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented:
Make it so, Jenkins.
